tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc

```diff
@@ -988,63 +988,55 @@ TEST_F(ConverterTest, GetTrtBroadcastShape) {
+                 expected_code, expected_error_msg_substr);
     if (expected_code == error::OK) {
       ExpectTrtDimsEqualsArray(expected_operand_1_shape, operand_1_new_dims);
       ExpectTrtDimsEqualsArray(expected_operand_2_shape, operand_2_new_dims);
     }
     // operand_2 broadcast operand_1
-    ExpectStatus(
-        this->converter_->GetTrtBroadcastShape(
-            operand_2, operand_1, &operand_2_new_dims, &operand_1_new_dims),
-        expected_code, expected_error_msg_substr);
+    ExpectStatus(GetTrtBroadcastShape(operand_2, operand_1, &operand_2_new_dims,
+                                      &operand_1_new_dims),
+                 expected_code, expected_error_msg_substr);
     if (expected_code == error::OK) {
       ExpectTrtDimsEqualsArray(expected_operand_1_shape, operand_1_new_dims);
       ExpectTrtDimsEqualsArray(expected_operand_2_shape, operand_2_new_dims);
     }
   };
 
-  // Both inputs are weights.
-  symmetric_test(
-      {1}, {1}, kIsNotTensor, kIsNotTensor, {}, {}, error::INVALID_ARGUMENT,
-      "Broadcasting requires at least one of the operands be tensors");
+  // Both inputs are weights. This should be handled by constfold grappler.
+  symmetric_test({1}, {1}, kIsNotTensor, kIsNotTensor, {1}, {1});
 
   // One tensor and one weights.
   symmetric_test({1, 1, 1}, {2}, kIsTensor, kIsNotTensor, {1, 1, 1}, {1, 1, 2});
   symmetric_test({1, 1, 2}, {2}, kIsTensor, kIsNotTensor, {1, 1, 2}, {1, 1, 2});
   symmetric_test({1, 3, 2}, {1}, kIsTensor, kIsNotTensor, {1, 3, 2}, {1, 1, 1});
   symmetric_test({1, 1, 1}, {2, 3}, kIsTensor, kIsNotTensor, {1, 1, 1},
                  {1, 2, 3});
+  symmetric_test({1, 1, 1}, {2, 3}, kIsTensor, kIsTensor, {1, 1, 1}, {1, 2, 3});
   symmetric_test({1, 1, 1}, {2, 3, 4}, kIsTensor, kIsNotTensor, {1, 1, 1},
                  {2, 3, 4});
   symmetric_test({1, 1, 1}, {1, 2, 3, 4}, kIsTensor, kIsNotTensor, {1, 1, 1},
                  {2, 3, 4});
   symmetric_test({1, 3, 4}, {1, 2, 1, 4}, kIsTensor, kIsNotTensor, {1, 3, 4},
                  {2, 1, 4});
   symmetric_test({1, 1, 1}, {2, 1, 1, 1}, kIsTensor, kIsNotTensor, {}, {},
-                 error::INVALID_ARGUMENT, "Infeasible broadcast scheme");
+                 error::INVALID_ARGUMENT,
+                 "Cannot broadcast weights with non-trivial batch dimension");
   symmetric_test({1, 1, 1}, {2, 1, 1, 1}, kIsTensor, kIsNotTensor, {}, {},
-                 error::INVALID_ARGUMENT, "Infeasible broadcast scheme",
-                 /*operand_1_batch_size=*/2);
-  symmetric_test({1, 1, 1}, {1, 1, 1, 1, 1}, kIsTensor, kIsNotTensor, {}, {},
                  error::INVALID_ARGUMENT,
-                 "Broadcasting beyond batch dimension is not supported "
-                 "(tensor #dims 4 vs broadcast #dims 5)");
+                 "Cannot broadcast weights with non-trivial batch dimension",
+                 /*operand_1_batch_size=*/2);
+  symmetric_test({1, 1, 1}, {1, 1, 1, 1, 1}, kIsTensor, kIsNotTensor,
```

Comment:
Copying comments from the other PR here:

> aaroey: Does it mean that elementwise layer broadcast ignores batch dimension? This is probably not what we want. Say we have a TF tensor A of shape [2,3], B of shape [1,1,3] (both include batch dim). Then output of tf.add(A, B) will have shape [1,2,3], but with ConvertBinary the output will have shape [2,1,3]? If that's not the case, please add such a test here (operand_1_batch_size would be 2 and operand_2_batch_size would be 1) and in TestBinaryTensorOpTensor.

> trevor-m: I don't think this is possible in TRT since all tensors must have the same batch size.

What I meant is, here the inputs are (tensor) shape [batch, 1, 1, 1] and (weights) shape [1, 1, 1, 1, 1], and the output is [batch, 1,1,1,1]. I'm wondering what the output would be if we change the input to be (tensor) shape [batch, 3] and (weights) shape [1, 1, 3]. Please add such a test and set batch to 2 (i.e. `operand_1_batch_size=2`), and make sure the output is __*not*__ [batch, 1, 3] which is __*not*__ valid in TF (while the expected output would be [1, batch, 3] which is then invalid in TRT, so the test should return error).

Reply:
Thanks for clarifying, yes it looks like the output is `[batch, 1, 3]` for the situation you described. I think this means we need to add a new requirement that (using TF definition of rank): `rank(weights) <= rank(tensor)`